src/Features/Core/Portable/NavigationBar/AbstractNavigationBarItemService.cs

```diff
@@ -35,15 +39,57 @@ public async Task<ImmutableArray<RoslynNavigationBarItem>> GetItemsAsync(Documen
             return items;
         }
 
-        protected static TextSpan? GetSelectionSpan(ISymbol symbol, SyntaxTree tree)
+        protected static SymbolItemLocation? GetSymbolLocation(
+            Solution solution, ISymbol symbol, SyntaxTree tree, Func<SyntaxReference, TextSpan> computeFullSpan)
         {
-            foreach (var location in symbol.Locations)
+            return GetSymbolLocation(solution, symbol, tree, computeFullSpan, symbol.DeclaringSyntaxReferences);
+        }
+
+        private static SymbolItemLocation? GetSymbolLocation(
+            Solution solution, ISymbol symbol, SyntaxTree tree,
+            Func<SyntaxReference, TextSpan> computeFullSpan,
+            ImmutableArray<SyntaxReference> allReferences)
+        {
+            if (allReferences.Length == 0)
+                return null;
+
+            // See if there are any references in the starting file. We always prefer those for any symbol we find.
+            var referencesInCurrentFile = allReferences.WhereAsArray(r => r.SyntaxTree == tree);
+            if (referencesInCurrentFile.Length > 0)
+            {
+                // the symbol had one or more declarations in this file.  We want to include all those spans in what we
+                // return so that if the use enters any of its spans we highlight it in the list.  An example of having
+                // multiple locations in the same file would be a a partial type with multiple parts in the same file.
+
+                // If we're not able to find a narrower navigation location in this file though then just navigate to
+                // the first reference itself.
+                var navigationLocationSpan = symbol.Locations.FirstOrDefault(loc => loc.SourceTree == tree)?.SourceSpan ??
+                                             referencesInCurrentFile.First().Span;
+
+                var spans = referencesInCurrentFile.SelectAsArray(r => computeFullSpan(r));
+                return new SymbolItemLocation((spans, navigationLocationSpan), otherDocumentInfo: null);
+            }
+            else
             {
-                if (location.SourceTree == tree)
-                    return location.SourceSpan;
+                // the symbol was defined in another file altogether.  We don't care about it's full span
+                // (since that is only needed for intersecting with the caret.  Instead, we just need a 
+                // reasonable location to navigate them to.  First try to find a narrow location to navigate to.
+                // And, if we can't, just go to the first reference we can find.
+                var navigationLocation = symbol.Locations.FirstOrDefault(loc => loc.SourceTree != null && loc.SourceTree != tree) ??
+                                         Location.Create(allReferences.First().SyntaxTree, allReferences.First().Span);
+
+                var documentId = solution.GetDocumentId(navigationLocation.SourceTree);
```

Comment:
should this be getrequireddocument?

Reply:
that's a good question.  i'll add a comment but this is just paranoia here.